cocina_level2_prs.rb

```diff
@@ -0,0 +1,185 @@
+GIT_BRANCH_FETCH_REFS = GIT_MAIN_FETCH_REFS.gsub('main', BRANCH_NAME).freeze
+COMMIT_DESCRIPTION = 'Update dependencies for cocina-models update'
+EXPECTED_PUSH_MESSAGE = "Branch '#{BRANCH_NAME}' set up to track remote branch '#{BRANCH_NAME}' from 'origin'"
+
+def repos_file
+  File.join 'infrastructure', 'projects.yml'
+end
+
+# @return [Array<String>] the git repos that need cocina level2 updates
+def repos
+  YAML.load_file(repos_file).fetch('projects')
+      .select { |project| project.fetch('cocina_level2', true) }
+      .map { |project| project.fetch('repo') }
+end
+
+def create_prs(repos)
+  repos.map do |repo|
+    pr_link = prepare_and_create_pr(repo)
+    p pr_link
+    { repo: repo, pr_link: pr_link }
+  end
+end
+
+def prepare_and_create_pr(repo)
+  clone_or_update_repo(repo)
+
+  within_cloned_repo_dir(repo) do
+    result = create_branch
+    result = update_gems unless result_failure?('create_branch', result)
+    result = local_git_commit unless result_failure?('update_gems', result)
+    result = git_push_origin unless result_failure?('local_git_commit', result)
+    create_pr if result.include?(EXPECTED_PUSH_MESSAGE)
+  end
+end
+
+# must be called as part of the block for within_cloned_repo_dir(repo)
+def create_branch
+  ErrorEmittingExecutor.execute("git checkout -B #{BRANCH_NAME}")
+
+  # Ensure local branch matches any existing upstream branch
+  ErrorEmittingExecutor.execute('git reset --hard', exit_on_error: true)
+end
+
+# must be called as part of the block for within_cloned_repo_dir(repo)
+def update_gems
+  gemfile = File.read('Gemfile')
+  %w[cocina-models dor-services-client sdr-client].each do |gem_name|
+    ErrorEmittingExecutor.execute("bundle update #{gem_name}") if gemfile.include?(gem_name)
+  end
+end
+
+# must be called as part of the block for within_cloned_repo_dir(repo)
+def local_git_commit
+  ErrorEmittingExecutor.execute("git commit -m '#{COMMIT_DESCRIPTION}' Gemfile.lock")
+end
+
+# must be called as part of the block for within_cloned_repo_dir(repo)
+def git_push_origin
+  ErrorEmittingExecutor.execute("git push origin -u #{BRANCH_NAME}")
+end
+
+# must be called as part of the block for within_cloned_repo_dir(repo)
+def create_pr
+  ErrorEmittingExecutor.execute("hub pull-request -f -m '#{COMMIT_DESCRIPTION}'")
+end
+
+# Check the result string of ErrorEmittingExecutor command for failure,
+#   output a message if so
+def result_failure?(step_name, result)
+  if result == 'FAILED'
+    p "stopped after #{step_name} due to failure"
+    return true
+  end
+  false
+end
+
+def clone_or_update_repo(repo)
+  puts "\n**** Processing #{repo} ****"
+  if File.exist?(cloned_repo_dir(repo))
+    update_repo(repo)
+  else
+    create_repo(repo)
+  end
+end
+
+def update_repo(repo)
+  within_cloned_repo_dir(repo) do
+    fetch_update
+    # we need local main branch to be up to date if we are creating a branch based on it
+    ErrorEmittingExecutor.execute('git switch main', exit_on_error: true)
```

Comment:
If we're fetching all branches and resetting the current branch to the current state of `origin/main`, I'm wondering if we need to switch into main at all.

Reply:
ok, let's not reset the current branch.  I'm sold.

Reply:
Cool, ditching this line